tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java

```diff
@@ -224,29 +230,53 @@ MediaType detectStreaming(InputStream input, Metadata metadata) throws IOExcepti
         } catch (Exception e) {
             //swallow
         }
-
+        
         return finalDetect(detectContext);
     }
-
-
+    
+    
     private MediaType detect(ZipArchiveEntry zae, ZipArchiveInputStream zis,
-                             StreamingDetectContext detectContext) throws IOException {
-        for (ZipContainerDetector d : zipDetectors) {
+            StreamingDetectContext detectContext) throws IOException {
+        for (ZipContainerDetector d : getDetectors()) {
             MediaType mt = d.streamingDetectUpdate(zae, zis, detectContext);
             if (mt != null) {
                 return mt;
             }
         }
         return null;
     }
-
+    
     private MediaType finalDetect(StreamingDetectContext detectContext) {
-        for (ZipContainerDetector d : zipDetectors) {
+        for (ZipContainerDetector d : getDetectors()) {
             MediaType mt = d.streamingDetectFinal(detectContext);
             if (mt != null) {
                 return mt;
             }
         }
         return MediaType.APPLICATION_ZIP;
     }
+    
+    private void setDetectors(List<ZipContainerDetector> zipDetectors) {
+        this.staticZipDetectors0 = zipDetectors;
+    }
+    
+    private List<ZipContainerDetector> getDetectors() {
+        ServiceLoader loader = getServiceLoader();
+        if (staticZipDetectors0 == null) {
+            staticZipDetectors0 = new ArrayList<ZipContainerDetector>();
+            staticZipDetectors0.addAll(loader.loadStaticServiceProviders(ZipContainerDetector.class));
+        }
+        List<ZipContainerDetector> zipDetectors = new ArrayList<>();
+        zipDetectors.addAll(staticZipDetectors0);
+        zipDetectors.addAll(loader.loadDynamicServiceProviders(ZipContainerDetector.class));
+        return zipDetectors;
+    }
+    
+    private ServiceLoader getServiceLoader() {
+        if (loader0 ==  null) {
+            loader0 = new ServiceLoader(DefaultZipContainerDetector.class.getClassLoader(), Boolean.getBoolean("org.apache.tika.service.error.warn") 
+                ? LoadErrorHandler.WARN:LoadErrorHandler.IGNORE, true);
+        }
+        return loader0;
+    }
```

Comment:
Is the 'DefaultZipContainerDetector' the only service that uses the service loader?

We should eagerly search for more potential errros, and fix them as well. Otherwise we just don't know whether we'll run into the same problem later again.

Reply:
there are many - I just would not touch something that works- Tika team apparently made it to not load dynamic services - so we would introduce new functionality by adding dynamic loading to all other classes

Reply:
We know that it does not work in OSGi.

How many places are 'many'?

Reply:
I count 8 - but the issue here is not that the service loader does not work - the issue is that they have not made the DefaultZipContainerDetector in a way to that it loads ZipContainerDetectors from other packages - so it also does not work otherwise i.e https://issues.apache.org/jira/browse/TIKA-3046

Reply:
I will add dynamic loading to the other classes